File: backend/app/schemas/dues.py
import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class SingleDuePayRequest(BaseModel):
    member_id: str = Field(..., description="Member UUID")
    month: int = Field(..., ge=1, le=12, description="Month (1-12)")
    year: int = Field(..., ge=2000, le=2100, description="Year (e.g. 2026)")
    amount: Optional[int] = Field(None, gt=0, description="Payment amount (defaults to configured monthly fee)")
    is_additional: bool = Field(False, description="Whether this is an additional extra contribution")
    payment_date: Optional[datetime.datetime] = Field(None, description="Payment timestamp (defaults to now)")
    payment_method: str = Field("CASH", description="Payment method: CASH, BANK, MOBILE_BANKING")
    reference_number: Optional[str] = Field(None, description="Optional payment reference / voucher number")
    notes: Optional[str] = Field(None, description="Optional remarks")


class MultiMonthDuePayRequest(BaseModel):
    member_id: str = Field(..., description="Member UUID")
    from_month: int = Field(..., ge=1, le=12, description="Starting month (1-12)")
    from_year: int = Field(..., ge=2000, le=2100, description="Starting year")
    to_month: int = Field(..., ge=1, le=12, description="Ending month (1-12)")
    to_year: int = Field(..., ge=2000, le=2100, description="Ending year")
    amount_per_month: Optional[int] = Field(None, gt=0, description="Fee per month (defaults to configured monthly fee)")
    payment_date: Optional[datetime.datetime] = Field(None, description="Payment timestamp (defaults to now)")
    payment_method: str = Field("CASH", description="Payment method: CASH, BANK, MOBILE_BANKING")
    reference_number: Optional[str] = Field(None, description="Optional payment reference / voucher number")
    notes: Optional[str] = Field(None, description="Optional remarks")


class PaidMonthDetail(BaseModel):
    month: int
    year: int
    amount: int
    monthly_contribution_id: str


class DuePaymentResponse(BaseModel):
    success: bool = True
    member_id: str
    total_amount_paid: int
    months_paid: List[PaidMonthDetail]
    paid_until_month: Optional[int] = None
    paid_until_year: Optional[int] = None
    ledger_transaction_id: str
    message: str


class MemberDuesSummaryResponse(BaseModel):
    member_id: str
    member_code: str
    full_name: str
    mobile: Optional[str] = None
    group_id: str
    group_name: str
    join_date: Optional[datetime.datetime] = None
    current_monthly_fee: int
    paid_until_month: Optional[int] = None
    paid_until_year: Optional[int] = None
    total_paid_amount: int
    total_due_amount: int
    status: str


class MemberDuesLedgerItem(BaseModel):
    id: str
    monthly_contribution_id: str
    month: int
    year: int
    expected_amount: int
    paid_amount: int
    is_additional: bool
    status: str
    payment_date: Optional[datetime.datetime] = None
    payment_method: Optional[str] = None
    reference_number: Optional[str] = None
    notes: Optional[str] = None
    ledger_transaction_id: Optional[str] = None


class MemberDuesLedgerResponse(BaseModel):
    member_id: str
    items: List[MemberDuesLedgerItem]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/lib/api/dues.ts
import { apiClient } from "./client";
import { PaginatedResult } from "./types";

export interface DueCalculationResponse {
  member_id: string;
  member_name: string;
  from_month: number;
  from_year: number;
  to_month: number;
  to_year: number;
  total_months: number;
  current_monthly_fee: number;
  total_amount: number;
  months_breakdown: Array<{
    month: number;
    year: number;
    amount: number;
    is_already_paid: boolean;
  }>;
}

export interface DuePaymentPayload {
  member_id: string;
  from_month: number;
  from_year: number;
  to_month: number;
  to_year: number;
  amount: number;
  payment_method?: string;
  date?: string;
  reference_number?: string | null;
  notes?: string | null;
}

export interface DuePaymentResponse {
  message: string;
  member_id: string;
  paid_months_count: number;
  total_amount: number;
  paid_until_month: number;
  paid_until_year: number;
  transaction_id: string;
}

export interface MemberDuesSummaryResponse {
  member_id: string;
  full_name: string;
  group_id: string;
  group_name: string;
  current_monthly_fee: number;
  paid_until_month: number;
  paid_until_year: number;
  total_paid_amount: number;
  total_paid_months: number;
}

export const duesApi = {
  async getFeeSetting(token?: string): Promise<{ monthly_fee: number; currency: string }> {
    return apiClient.get<{ monthly_fee: number; currency: string }>("/api/v1/dues/settings/fee", { token });
  },

  async calculate(params: {
    member_id: string;
    from_month: number;
    from_year: number;
    to_month: number;
    to_year: number;
  }, token?: string): Promise<DueCalculationResponse> {
    return apiClient.get<DueCalculationResponse>("/api/v1/dues/calculate", { params, token });
  },

  async pay(data: DuePaymentPayload, token?: string): Promise<DuePaymentResponse> {
    return apiClient.post<DuePaymentResponse>("/api/v1/dues/pay", data, { token });
  },

  async getSummary(memberId: string, token?: string): Promise<MemberDuesSummaryResponse> {
    return apiClient.get<MemberDuesSummaryResponse>("/api/v1/dues/summary", { params: { member_id: memberId }, token });
  },
};
